plugins/kubernetes/app/models/watchers/deploy_watcher.rb

```diff
@@ -7,113 +6,259 @@ class DeployWatcher
 
     finalizer :on_termination
 
-    def initialize(release)
-      @release = release
-      @current_rcs = {}
+    def initialize(project)
+      @project = project
       @pod_timer = after(ENV.fetch('KUBERNETES_POD_TIMEOUT', 10.minutes).to_i) { pod_timeout }
-      info "Start watching K8s deploy: #{@release}"
+      start_watching
+    end
+
+    def self.watcher_symbol(project)
+      "deploy-watcher-#{project.id}".to_sym
+    end
+
+    def self.start_watcher(project)
+      watcher_name = watcher_symbol(project)
+      supervise as: watcher_name, args: [project]
+    end
+
+    def self.restart_watcher(project)
+      watcher = Actor[watcher_symbol(project)]
+      watcher.terminate if watcher and watcher.alive?
+      start_watcher(project)
+    end
+
+    def start_watching
+      sync_with_cluster
+      info "Start watching deployment for project: #{project.name}"
       async :watch
     end
 
     def watch
-      subscribe("#{@release.build.project_name}", :handle_update)
-    end
-
-    def handle_update(topic, data)
-      release_doc = release_doc_from_rc_name(topic)
-      release_updated = if data.object.kind == 'Event'
-                        handle_event_update(release_doc)
-                      else
-                        handle_pod_update(release_doc, data)
-                      end
-      if release_updated
-        send_event(role: release_doc.kubernetes_role.name,
-                   deploy_group: release_doc.deploy_group.name,
-                   target_replicas: release_doc.replica_target,
-                   live_replicas: release_doc.replicas_live,
-                   failed: release_doc.failed?)
-        end_deploy if deploy_finished?
+      subscribe(Watchers::TopicSubscription.pod_updates_topic(project.id), :handle_event)
+    end
+
+    def handle_event(topic, message)
+      event = message.event
+      debug "Got message on topic: #{topic}. Type of message: #{event.kind}"
+
+      case
+      when event.kind == 'Event' then handle_cluster_event(message)
+      when event.kind == 'Pod' then handle_pod_event(message)
+      else error "Object kind not yet supported: #{notice.object.kind}"
       end
     end
 
     private
 
-    def handle_event_update(release_doc)
-      release_doc.failed? ? false : release_doc.fail!
+    attr_reader :project
```

Comment:
Again not really seeing the point behind private accessors instead of just using ```@project```